File: December 2022/Day 3/JAVA Code/Aggressive Cows.java
// Problem Link :- https://practice.geeksforgeeks.org/problems/aggressive-cows/1
// Video Solution Link :- 

//{ Driver Code Starts
// Initial Template for Java

import java.io.*;
import java.util.*;

class GFG {
    public static void main(String[] args) throws IOException {
        Scanner sc = new Scanner(System.in);
        int t;
        t = sc.nextInt();
        while (t-- > 0) {

            int n;
            n = sc.nextInt();

            int k;
            k = sc.nextInt();

            int[] v = new int[n];
            for (int i = 0; i < n; i++) v[i] = sc.nextInt();

            Solution obj = new Solution();
            int res = obj.solve(n, k, v);

            System.out.println(res);
        }
    }
}

// } Driver Code Ends


// User function Template for Java
class Solution {
    // Time Complexity :- O(n.log(10^9));
    // Space Complexity :- O(1);
    
    public static int solve(int n, int k, int[] stalls) {
        Arrays.sort(stalls);
        int lo=0,hi=stalls[stalls.length-1];
        int ans=0;
        while(lo<=hi) {
            int mid = lo+(hi-lo)/2;
            if(isValid(stalls,k,mid)) {
                ans = Math.max(ans,mid);
                lo=mid+1;
            } else {
                hi=mid-1;
            }
        }
        return ans;
    }
    
    public static boolean isValid(int[] a,int k,int dis) {
        int pidx=0;
        for(int i=1;i<a.length;i++) {
            if(a[i]-a[pidx]>=dis) {
                k--;
                pidx=i;
            }
        }
        return k<=1;
    }
    
}

File: December 2022/Day 3/C++ Code/Aggressive Cows.cpp
// Problem Link :- https://practice.geeksforgeeks.org/problems/aggressive-cows/1
// Video Solution Link :- https://youtu.be/7OvKCVz79sc

//{ Driver Code Starts
// Initial Template for C++
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
// User function Template for C++

class Solution {
public:
    // Time Complexity :- O(n.log(10^9));
    // Space Complexity :- O(1);
    
    int solve(int n, int k, vector<int> &stalls) {
    
        // Write your code here
        sort(stalls.begin(),stalls.end());
        int lo=0,hi=stalls[stalls.size()-1];
        int ans=0;
        while(lo<=hi) {
            int mid = lo+(hi-lo)/2;
            if(isValid(stalls,k,mid)) {
                ans = max(ans,mid);
                lo=mid+1;
            } else {
                hi=mid-1;
            }
        }
        return ans;
    }
    
    bool isValid(vector<int> a,int k,int dis) {
        int pidx=0;
        for(int i=1;i<a.size();i++) {
            if(a[i]-a[pidx]>=dis) {
                k--;
                pidx=i;
            }
        }
        return k<=1;
    }
};

//{ Driver Code Starts.

int main() {
    int t = 1;
    cin >> t;

    // freopen ("output_gfg.txt", "w", stdout);

    while (t--) {
        // Input

        int n, k;
        cin >> n >> k;

        vector<int> stalls(n);
        for (int i = 0; i < n; i++) {
            cin >> stalls[i];
        }
        // char ch;
        // cin >> ch;

        Solution obj;
        cout << obj.solve(n, k, stalls) << endl;

        // cout << "~\n";
    }
    // fclose(stdout);

    return 0;
}
// } Driver Code Ends
